src/analysis/normed_space/hahn_banach/separation.lean

```diff
@@ -0,0 +1,342 @@
+  rintro _ ⟨_, _, rfl⟩,
+  exact hf₂ _ ‹_›,
+end
+
+open filter
+open_locale topological_space
+
+/-- If `A`, `B` are disjoint convex sets, `A` is compact and `B` is closed then we can find open
+disjoint convex sets containing them. -/
+-- TODO: This proof uses the normed space structure of `E`, but it could work for locally convex
+-- topological vector spaces: instead of taking the balls around 0 with radius 1/n, we could show
+-- there must be some convex neighbourhood `W` of 0 which make `A + W` and `B + W` disjoint?
+theorem closed_compact_separate {A B : set E} (hA₁ : convex ℝ A) (hA₂ : is_compact A)
+  (hB₁ : convex ℝ B) (hB₃ : is_closed B) (disj : disjoint A B) :
+  ∃ U V, is_open U ∧ is_open V ∧ convex ℝ U ∧ convex ℝ V ∧ A ⊆ U ∧ B ⊆ V ∧ disjoint U V :=
+begin
+  have : ∃ (n : ℕ), disjoint (A + metric.ball 0 (n+1)⁻¹) (B + metric.ball 0 (n+1)⁻¹),
+  { by_contra h,
+    push_neg at h,
+    simp only [not_disjoint_iff, set.mem_add, metric.mem_ball, dist_zero_right,
+      ←exists_and_distrib_left, ←exists_and_distrib_right, and_assoc] at h,
+    choose z f f' g g' h₁ h₂ h₃ h₄ h₅ h₆ using h,
+    obtain ⟨w, hw, φ, hφ₁, hφ₂ : tendsto (f ∘ _) _ _⟩ := hA₂.tendsto_subseq h₁,
+    have : tendsto (g ∘ φ) at_top (𝓝 w),
+    { have : tendsto (f - g) at_top (𝓝 0),
+      { suffices : ∀ n, ∥(f - g) n∥ ≤ 2 * (n+1)⁻¹,
+        { apply squeeze_zero_norm this,
+          rw ←mul_zero (2:ℝ),
+          apply tendsto.const_mul (2:ℝ),
+          simp_rw inv_eq_one_div,
+          apply tendsto_one_div_add_at_top_nhds_0_nat },
+        intro n,
+        simp only [pi.sub_apply],
+        have : f n - g n = g' n - f' n,
+        { rw [sub_eq_iff_eq_add', ←add_sub_assoc, h₆, ←h₃, add_sub_cancel] },
+        rw this,
+        apply le_trans (norm_sub_le _ _) _,
+        rw two_mul,
+        apply add_le_add (h₅ n).le (h₂ n).le },
+      have : tendsto (f ∘ φ - g ∘ φ) at_top (𝓝 0),
+      { have : f ∘ φ - g ∘ φ = (f - g) ∘ φ,
+        { ext,
+          simp },
+        rw this,
+        apply tendsto.comp ‹tendsto (f - g) at_top _› hφ₁.tendsto_at_top },
+      simpa using tendsto.sub hφ₂ ‹tendsto (f ∘ φ - g ∘ φ) at_top _› },
+    have := mem_of_is_closed_sequential ‹is_closed B› (λ n, h₄ (φ n)) this,
+    apply disj ⟨hw, ‹w ∈ B›⟩ },
+  obtain ⟨n, hn⟩ := this,
+  refine ⟨_, _, _, _, hA₁.add _, hB₁.add _, _, _, hn⟩,
+  { exact metric.is_open_ball.add_left },
+  { exact metric.is_open_ball.add_left },
+  { exact convex_ball 0 _ },
+  { exact convex_ball 0 _ },
+  { suffices : A + {0} ⊆ A + metric.ball (0:E) (n+1)⁻¹,
+    { simpa },
+    apply add_subset_add (set.subset.refl _),
+    simp only [metric.mem_ball, norm_zero, dist_zero_left, singleton_subset_iff, inv_pos],
+    norm_cast,
+    simp },
+  { suffices : B + {0} ⊆ B + metric.ball (0:E) (n+1)⁻¹,
+    { simpa },
+    apply add_subset_add (set.subset.refl _),
+    simp only [metric.mem_ball, norm_zero, dist_zero_left, singleton_subset_iff, inv_pos],
+    norm_cast,
+    simp },
```

Comment:
More golf.
```suggestion
  refine ⟨_, _, metric.is_open_ball.add_left, metric.is_open_ball.add_left,
    hA₁.add (convex_ball 0 _), hB₁.add (convex_ball 0 _), _, _, hn⟩,
  { refine subset_add_right A _ (metric.mem_ball_self _),
    sorry, },
  { refine subset_add_right B _ (metric.mem_ball_self _),
    sorry, },
```
The two `sorry` are `0 < (↑n + 1)⁻¹`. This code needs the following new lemma, to be placed somewhere appropriate:
```lean
@[to_additive]
lemma subset_mul_right {α} [monoid α] (s t : set α) (ht : (1 : α) ∈ t) : s ⊆ s * t :=
begin
  suffices : s * {1} ⊆ s * t, by simpa only [image_id', mul_one, mul_singleton],
  apply mul_subset_mul (set.subset.refl _),
  simpa,
end
```

Reply:
Wow, very nice!